pkg/cli/cli.go

```diff
@@ -0,0 +1,276 @@
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold/project"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	noticeColor = "\033[1;36m%s\033[0m"
+)
+
+// Option is a function that can configure the CLI
+type Option func(*CLI) error
+
+// CLI defines the command line structure and interfaces that are used to
+// scaffold kubebuilder project files.
+type CLI struct {
+	commandName    string
+	defProjVersion string
+	cmd            *cobra.Command
+	extraCommands  []*cobra.Command
+	plugins        map[string][]plugin.Base
+}
+
+// New creates a new CLI instance.
+func New(opts ...Option) (*CLI, error) {
+	c := &CLI{
+		commandName:    "kubebuilder",
+		defProjVersion: project.Version2,
+		plugins:        map[string][]plugin.Base{},
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := c.initialize(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run runs the CLI.
+func (c CLI) Run() error {
+	return c.cmd.Execute()
+}
+
+// WithCommandName is an Option that sets the CLI's root command name.
+func WithCommandName(name string) Option {
+	return func(c *CLI) error {
+		c.commandName = name
+		return nil
+	}
+}
+
+// WithDefaultProjectVersion is an Option that sets the CLI's default project
+// version. Setting an unknown version will result in an error.
+func WithDefaultProjectVersion(version string) Option {
+	return func(c *CLI) error {
+		c.defProjVersion = version
+		return nil
+	}
+}
+
+// WithPlugins is an Option that sets the CLI's plugins.
+func WithPlugins(plugins ...plugin.Base) Option {
+	return func(c *CLI) error {
+		for _, p := range plugins {
+			ver := p.Version()
+			if _, ok := c.plugins[ver]; !ok {
+				c.plugins[ver] = []plugin.Base{}
+			}
+			c.plugins[ver] = append(c.plugins[ver], p)
+		}
+		return nil
+	}
+}
+
+// WithExtraCommands is an Option that adds extra subcommands to the CLI.
+// Adding extra commands that duplicate existing commands results in an error.
+func WithExtraCommands(cmds ...*cobra.Command) Option {
+	return func(c *CLI) error {
+		c.extraCommands = cmds
+		return nil
+	}
+}
+
+// initialize initializes the CLI.
+func (c *CLI) initialize() error {
+	// Inject the command name into each plugin that supports
+	// injectable command names.
+	for _, ps := range c.plugins {
+		for _, p := range ps {
+			if in, ok := p.(plugin.InjectsCommandName); ok {
+				in.InjectCommandName(c.commandName)
+			}
+		}
+	}
+
+	// Check the current working directory for a PROJECT file.
+	// Return whether it was found, and if so, what project version
+	// it is for.
+	foundProject, projectVersion := getProjectVersion()
+
+	// If a PROJECT file wasn't found, see if we can find the
+	// project version in a command line flag.
+	if !foundProject {
+		projectVersion, _ = c.getBaseFlags()
```

Comment:
As per my other comment, I think we can actually just move project-version to `init`.  That still leaves plugin though.

Reply:
Handling this information will be easier if we merge #1325 first.

Reply:
The  #1325 is merged now @estroz